src/System.Text.Primitives/System/Text/Encoding/Utf16/Utf16LittleEndianEncoder.cs

```diff
@@ -2,102 +2,204 @@
-                encodedBytes = default(int);
-                return false;
+                int written;
+                if (!TryEncode(utf32[i], utf16, charactersWritten, out written))
+                    return false;
+
+                charactersConsumed++;
+                charactersWritten += written;
             }
 
-            encodedBytes = UnicodeCodePoint.IsBmp(codePoint) ? 2 : 4;
+            return true;
+        }
 
-            if (buffer.Length < encodedBytes)
+        /// <summary>
+        /// Encodes a single UTF-32 character into UTF-16.
+        /// </summary>
+        /// <param name="codePoint">A UTF-32 character to encode.</param>
+        /// <param name="utf16">A span to write the UTF-16 encoded data into.</param>
+        /// <param name="index">An index into the <paramref name="utf16"/> span to to the write operation.</param>
+        /// <param name="charactersWritten">An output parameter to store the number of bytes written to <paramref name="utf16"/></param>
+        /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
+        public static bool TryEncode(uint codePoint, Span<char> utf16, int index, out int charactersWritten)
+        {
+            if (!UnicodeHelpers.IsSupportedCodePoint(codePoint))
             {
-                codePoint = default(UnicodeCodePoint);
-                encodedBytes = default(int);
-                // buffer too small
+                charactersWritten = 0;
                 return false;
             }
 
-            if (encodedBytes == 2)
+            charactersWritten = UnicodeHelpers.IsBmp(codePoint) ? 1 : 2;
+
+            if (utf16.Length - index < charactersWritten)
             {
-                unchecked
-                {
-                    buffer.Write((ushort)codePoint);
-                }
+                charactersWritten = 0;
+                return false;
             }
-            else
+
+            unchecked
             {
-                unchecked
+                if (charactersWritten == 1)
+                    utf16[index] = (char)codePoint;
+                else
                 {
-                    uint codePointValue = (uint)codePoint;
-                    uint highSurrogate = ((codePointValue - 0x010000u) >> 10) + UnicodeConstants.Utf16HighSurrogateFirstCodePoint;
-                    uint lowSurrogate = (codePointValue & MaskLow10Bits) + UnicodeConstants.Utf16LowSurrogateFirstCodePoint;
-                    buffer.Write(highSurrogate | (lowSurrogate << 16));
+                    utf16[index] = (char)(((codePoint - 0x010000u) >> 10) + Utf16HighSurrogateFirstCodePoint);
+                    utf16[index + 1] = (char)((codePoint & 0x3FF) + Utf16LowSurrogateFirstCodePoint);
                 }
             }
+
             return true;
         }
 
+        #endregion Encoding implementation
+
+        #region Surrogate helpers
+
+        private static bool IsSurrogate(uint codePoint)
```

Comment:
We should think about it. We don't want to add APIs that will not be used, but on the other hand there is risk of people reimplementing these all the time.